api/src/main/scala/za/co/absa/loginsvc/rest/provider/kerberos/KerberosSPNEGOAuthentication.scala

```diff
@@ -0,0 +1,122 @@
+    ticketValidator.setDebug(kerberosDebug)
+    ticketValidator.afterPropertiesSet()
+    ticketValidator
+  }
+
+  private def loginConfig(): SunJaasKrb5LoginConfig = {
+    val loginConfig = new SunJaasKrb5LoginConfig()
+    loginConfig.setServicePrincipal("SVC-CPS-LOGIN-LDAP@CORP.DSARENA.COM")
+    loginConfig.setKeyTabLocation(new FileSystemResource(kerberos.keytabFileLocation))
+    loginConfig.setDebug(kerberosDebug)
+    loginConfig.setIsInitiator(true)
+    loginConfig.setUseTicketCache(false)
+    loginConfig.afterPropertiesSet()
+    loginConfig
+  }
+
+  private def kerberosLdapContextSource(): KerberosLdapContextSource = {
+    val contextSource = new KerberosLdapContextSource(activeDirectoryLDAPConfig.url)
+    contextSource.setLoginConfig(loginConfig())
+    contextSource.afterPropertiesSet()
+    contextSource
+  }
+
+  private def ldapUserDetailsService(): LdapUserDetailsService = {
+    val userSearch = new kerberosLdapSearch(activeDirectoryLDAPConfig.domain, activeDirectoryLDAPConfig.searchFilter, kerberosLdapContextSource())
+    val service = new LdapUserDetailsService(userSearch, new ActiveDirectoryLdapAuthoritiesPopulator)
+    service.setUserDetailsMapper(new LdapUserDetailsMapper())
+    service
+  }
+
+  def kerberosServiceAuthenticationProvider(): KerberosServiceAuthenticationProvider = {
+    val provider = new KerberosServiceAuthenticationProvider()
+    provider.setTicketValidator(sunJaasKerberosTicketValidator())
+    provider.setUserDetailsService(ldapUserDetailsService())
+    provider
+  }
+}
+
+object RestApiKerberosAuthentication {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  def spnegoAuthenticationProcessingFilter(authenticationManager: AuthenticationManager): SpnegoAuthenticationProcessingFilter = {
+    val filter = new SpnegoAuthenticationProcessingFilter()
+    filter.setAuthenticationManager(authenticationManager)
+    filter.setSkipIfAlreadyAuthenticated(true)
+    filter
+  }
+}
+
+private class kerberosLdapSearch(searchBase: String, searchFilter: String, contextSource: BaseLdapPathContextSource)
+  extends FilterBasedLdapUserSearch(searchBase, searchFilter, contextSource) {
+  override def searchForUser(username: String): DirContextOperations = {
+    val user = if(username.contains("@")) {
+      username.split("@").head
+    }
+    else {
+      username
+    }
+    super.searchForUser(user)
+  }
+}
+
+private class ActiveDirectoryLdapAuthoritiesPopulator() extends LdapAuthoritiesPopulator {
+  override def getGrantedAuthorities(userData: DirContextOperations, username: String): util.Collection[_ <: GrantedAuthority] = {
+    val groups = userData.getStringAttributes("memberof")
+
+    if(groups.isEmpty)
+      AuthorityUtils.NO_AUTHORITIES
+    else {
+      groups.map({group =>
+        val ldapName = new LdapName(group)
+        val role = ldapName.getRdn(ldapName.size()-1).getValue.toString
+        new SimpleGrantedAuthority(role)
+      }).toList.asJava
+    }
+  }
```

Comment:
What is the point of this piece of code? How does it work?
Users can have bunch of `memberOf` records on `User` node.
Does it extract the the simple name of the member group from e.g. `CN=PAM-PROD-ZA-DEVOPS-USERS,OU=Security,OU=Groups,OU=CORP Accounts,DC=corp,DC=dsarena,DC=com` ? The part that confuses me the most is `ldapName.getRdn(ldapName.size()-1).getValue.toString`

Reply:
Yes, as far as I can see is that this is a way of handling Ldap responses and the code above is how to find the relevant information we need to grant the authority. I was just trying to test this as this how enceladus handles it, so wanted to see if this is a better method. My Testing currently doesn't get this far to see if this is correct or not yet though, unfortunately.